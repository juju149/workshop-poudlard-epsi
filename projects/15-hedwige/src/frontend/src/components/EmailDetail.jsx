import React from 'react';
import '../styles/EmailDetail.css';

function EmailDetail({ email, onBack }) {
  if (!email) {
    return <div className="email-detail-empty">Select an email to read</div>;
  }

  const formatDate = (dateString) => {
    const date = new Date(dateString);
    return date.toLocaleString('en-US', {
      weekday: 'short',
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    });
  };

  const extractEmail = (from) => {
    const match = from.match(/<(.+)>/);
    return match ? match[1] : from;
  };

  const extractName = (from) => {
    const match = from.match(/^(.+?)\s*</);
    return match ? match[1].replace(/"/g, '') : extractEmail(from);
  };

  return (
    <div className="email-detail">
      <div className="email-detail-header">
        <button className="back-btn" onClick={onBack}>
          ‚Üê Back to Inbox
        </button>
      </div>

      <div className="email-detail-content">
        <h2 className="email-detail-subject">{email.subject || '(No Subject)'}</h2>
        
        <div className="email-detail-meta">
          <div className="email-from-info">
            <strong>{extractName(email.from)}</strong>
            <span className="email-address">&lt;{extractEmail(email.from)}&gt;</span>
          </div>
          <div className="email-to-info">
            <span>To: {email.to}</span>
          </div>
          <div className="email-date-info">
            {formatDate(email.date)}
          </div>
        </div>

        <div className="email-body">
          <pre>{email.body || email.snippet}</pre>
        </div>
      </div>
    </div>
  );
}

export default EmailDetail;
