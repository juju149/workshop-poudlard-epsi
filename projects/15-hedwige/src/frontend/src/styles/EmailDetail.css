.email-detail {
  padding: 20px;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.email-detail-header {
  margin-bottom: 20px;
}

.back-btn {
  padding: 8px 16px;
  background: #f0f0f0;
  border: none;
  border-radius: 6px;
  color: #333;
  font-weight: 500;
  transition: background 0.2s;
}

.back-btn:hover {
  background: #e0e0e0;
}

.email-detail-content {
  flex: 1;
  overflow-y: auto;
}

.email-detail-subject {
  font-size: 1.8em;
  color: #333;
  margin-bottom: 20px;
  word-wrap: break-word;
}

.email-detail-meta {
  padding: 15px 0;
  border-bottom: 1px solid #e0e0e0;
  margin-bottom: 30px;
}

.email-from-info {
  margin-bottom: 8px;
}

.email-from-info strong {
  color: #333;
  font-size: 1.05em;
}

.email-address {
  color: #666;
  font-size: 0.9em;
  margin-left: 8px;
}

.email-to-info,
.email-date-info {
  color: #666;
  font-size: 0.9em;
  margin-top: 4px;
}

.email-body {
  line-height: 1.6;
  color: #333;
}

.email-body pre {
  white-space: pre-wrap;
  word-wrap: break-word;
  font-family: inherit;
  font-size: 1em;
}

.email-detail-empty {
  padding: 60px 20px;
  text-align: center;
  color: #999;
  font-size: 1.1em;
}
