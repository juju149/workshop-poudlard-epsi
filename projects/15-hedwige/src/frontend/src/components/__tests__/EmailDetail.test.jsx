import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EmailDetail from '../EmailDetail';
import { vi } from 'vitest';

describe('EmailDetail', () => {
  const email = {
    id: 1,
    from: 'Minerva McGonagall <mcgonagall@poudlard.fr>',
    to: 'Harry Potter <harry@poudlard.fr>',
    subject: 'Bienvenue à Poudlard',
    date: '2025-10-10T10:00:00Z',
    body: 'Cher Harry, bienvenue à Poudlard!'
  };

  it('affiche le message si aucun email n’est sélectionné', () => {
    render(<EmailDetail email={null} onBack={() => {}} />);
    expect(screen.getByText(/Select an email to read/i)).toBeInTheDocument();
  });

  it('affiche les informations de l’email', () => {
    render(<EmailDetail email={email} onBack={() => {}} />);
    // Vérifie que le sujet est bien affiché dans le h2
    const subjects = screen.getAllByText(/Bienvenue à Poudlard/i);
    expect(subjects[0].tagName).toBe('H2');
    expect(screen.getByText(/Minerva McGonagall/i)).toBeInTheDocument();
    expect(screen.getByText(/mcgonagall@poudlard.fr/i)).toBeInTheDocument();
    expect(screen.getByText(/To: Harry Potter <harry@poudlard.fr>/i)).toBeInTheDocument();
    expect(screen.getByText(/Back to Inbox/i)).toBeInTheDocument();
  });

  it('appelle onBack lors du clic sur le bouton', () => {
    const handleBack = vi.fn();
    render(<EmailDetail email={email} onBack={handleBack} />);
    fireEvent.click(screen.getByText(/Back to Inbox/i));
    expect(handleBack).toHaveBeenCalled();
  });

  it('correspond au snapshot', () => {
    const { asFragment } = render(<EmailDetail email={email} onBack={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

