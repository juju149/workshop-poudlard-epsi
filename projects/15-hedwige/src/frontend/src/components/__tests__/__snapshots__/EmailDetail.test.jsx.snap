// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailDetail > correspond au snapshot 1`] = `
<DocumentFragment>
  <div
    class="email-detail"
  >
    <div
      class="email-detail-header"
    >
      <button
        class="back-btn"
      >
        ← Back to Inbox
      </button>
    </div>
    <div
      class="email-detail-content"
    >
      <h2
        class="email-detail-subject"
      >
        Bienvenue à Poudlard
      </h2>
      <div
        class="email-detail-meta"
      >
        <div
          class="email-from-info"
        >
          <strong>
            Minerva McGonagall
          </strong>
          <span
            class="email-address"
          >
            &lt;mcgonagall@poudlard.fr&gt;
          </span>
        </div>
        <div
          class="email-to-info"
        >
          <span>
            To: Harry Potter &lt;harry@poudlard.fr&gt;
          </span>
        </div>
        <div
          class="email-date-info"
        >
          Fri, Oct 10, 2025, 12:00 PM
        </div>
      </div>
      <div
        class="email-body"
      >
        <pre>
          Cher Harry, bienvenue à Poudlard!
        </pre>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
